Create new appimage workflow

Jammyjamjamman opened the pull request:
Supports more architectures + means some workflow jobs can be removed.

Jammyjamjamman committed:
Create new appimage workflow

Supports more architectures + means some workflow jobs can be removed.

andy5995 committed:
Don't use 'out' directory (#279)

Jammyjamjamman committed:
remove "cd out" from appimage.yml

andy5995 committed:
Generate multiple checksums (#280)

Jammyjamjamman committed:
Fix artifact naming bug in appimage.yml

andy5995 commented:
As for the failing action on arm/v7, you could use curl instead of wget. I use this in the linuxdeploy container dockerfile:  `curl -LO https://raw.githubusercontent.com/linuxdeploy/linuxdeploy-plugin-gtk/3b67a1d1c1b0c8268f57f2bce40fe2d33d409cea/linuxdeploy-plugin-gtk.sh`

I have no idea why the amd64 build had no problem, but that one does.

andy5995 committed:
Use curl, and more variables (#281)

andy5995 committed:
Set install prefix, prevent duplicates (#282)

* Set install prefix, prevent duplicates

* Exit if VERSION is not set

andy5995 committed:
Use andy5995/linuxdeploy:v2 instead of ':latest'

Jammyjamjamman committed:
Add data to megaglest-game appimages

andy5995 committed:
appimage.yml: Use v2-focal instead of v2

andy5995 committed:
appimage.yml: Assign version based on workflow dispatch input

Use "snapshot" if input is not given

andy5995 committed:
Package all 3 binaries

andy5995 committed:
build-appimage.sh: Edit test instructions [skip ci]

andy5995 committed:
Remove SOURCE_ROOT variable and experimental shared library arg

andy5995 committed:
Actually use the docker-compose.yml added in previous commit

andy5995 committed:
Add comments to compose yml and .env, move vars from .env to compose

andy5995 committed:
Don't bother setting WORKSPACE in the script

andy5995 committed:
Don't build on arm/v7

andy5995 committed:
Install far few packages with --no-install-recommends

andy5995 committed:
Use v3 instead of v2, remove entrypoint

Using it here overrides the default set in the linuxdeploy image

andy5995 commented:
Arm64 build is now failing due to a segfault.

```
[ 60%] Built target libmegaglest
[ 60%] Building CXX object source/glest_map_editor/CMakeFiles/megaglest_editor.dir/main.cpp.o
[ 60%] Building CXX object source/glest_map_editor/CMakeFiles/megaglest_editor.dir/__/glest_game/global/config.cpp.o
[ 61%] Building CXX object source/g3d_viewer/CMakeFiles/megaglest_g3dviewer.dir/__/glest_game/graphics/particle_type.cpp.o
[ 61%] Building CXX object source/glest_game/CMakeFiles/megaglest.dir/ai/ai.cpp.o
c++: internal compiler error: Segmentation fault signal terminated program cc1plus
```

This could mean some Ubuntu focal package has changed since the last successful build, or something related to the qemu action (I see they did a minor release (v3.3.0) two weeks ago); or it might be something else entirely. An unknown element. A variable that doesn't not exist in our universe but at the quantum level.

https://github.com/docker/setup-qemu-action/tags

I'll specify the v3.2.0 tag and see if the results change.

andy5995 commented on .github/workflows/appimage.yml:
```diff
@@ -0,0 +1,74 @@
+
+
+      - if: ${{ ! contains(matrix.platform, 'amd64') }}
+        uses: docker/setup-qemu-action@v3
```
```suggestion
        uses: docker/setup-qemu-action@v3.2.0
```

Note 'v3' points to v3.3.0 https://github.com/docker/setup-qemu-action/tags This is that standard way github actions are released. When minor versions are released, the major tag is updated on the action repo.

andy5995 committed:
Specify v3.2.0 for qemu action

andy5995 commented:
After changing the qemu version to v3.2.0 the build fails with the same error at 12%

andy5995 commented on .github/workflows/appimage.yml:
```diff
@@ -0,0 +1,74 @@
+
+
+      - if: ${{ ! contains(matrix.platform, 'amd64') }}
+        uses: docker/setup-qemu-action@v3.2.0
```
```suggestion
        uses: docker/setup-qemu-action@v3
```

andy5995 committed:
Revert cb3bbcc

Jammyjamjamman committed:
Add CC and CXX variables docker-compose

Also switch appimage workflow to clang for now (gcc breaks arm64 build)

andy5995 commented:
After switching to clang, different error now

```

linuxdeploy -d $GAME_DESKTOP_DEST/megaglest.desktop \
  --icon-file=megaglest.png \
  --icon-filename=megaglest \
  --custom-apprun="$WORKSPACE/mk/linux/AppRun" \
  --executable AppDir/$INST_PREFIX/bin/megaglest \
  --appdir AppDir \
  --plugin gtk \
  --output appimage
linuxdeploy version 1-alpha (git commit ID 97f3b52), <local dev build> built on 2024-12-29 01:06:10 UTC

-- Creating basic AppDir structure -- 
Creating directory AppDir/usr/bin/ 
Creating directory AppDir/usr/lib/ 
Creating directory AppDir/usr/share/applications/ 
Creating directory AppDir/usr/share/icons/hicolor/ 
Creating directory AppDir/usr/share/icons/hicolor/16x16/apps/ 
Creating directory AppDir/usr/share/icons/hicolor/32x32/apps/ 
Creating directory AppDir/usr/share/icons/hicolor/64x64/apps/ 
Creating directory AppDir/usr/share/icons/hicolor/128x128/apps/ 
Creating directory AppDir/usr/share/icons/hicolor/256x256/apps/ 
Creating directory AppDir/usr/share/icons/hicolor/scalable/apps/ 

-- Deploying dependencies for existing files in AppDir -- 
Deploying dependencies for ELF file AppDir/usr/bin/megaglest 
terminate called after throwing an instance of 'std::runtime_error'
  what():  Failed to run ldd: exited with code 1
Aborted (core dumped)
```

andy5995 commented:
Retriggered with no changes but got a different error:

```
[ 30%] Building C object source/shared_lib/CMakeFiles/libmegaglest.dir/sources/feathery_ftp/ftpTargetPosix.c.o
[ 31%] Building C object source/shared_lib/CMakeFiles/libmegaglest.dir/sources/feathery_ftp/ftpTargetWin32.c.o
Segmentation fault (core dumped)
make[2]: *** [source/shared_lib/CMakeFiles/libmegaglest.dir/build.make:205: source/shared_lib/CMakeFiles/libmegaglest.dir/sources/feathery_ftp/ftpTargetWin32.c.o] Error 139
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/Makefile2:269: source/shared_lib/CMakeFiles/libmegaglest.dir/all] Error 2
make: *** [Makefile:156: all] Error 2
```

andy5995 commented on mk/linux/docker/docker-compose.yml:
```diff
@@ -0,0 +1,21 @@
+      WORKSPACE: ${WORKSPACE}
+      CC: "${CC:-gcc}"
+      CXX: "${CXX:-g++}"
+    image: andy5995/linuxdeploy:v3-focal
```
```suggestion
    image: andy5995/linuxdeploy:v3-jammy
```

andy5995 committed:
Use jammy image instead of focal

andy5995 commented:
After trying in the jammy container:

```
In file included from /workspace/source/glest_game/world/scenario.h:24:
/workspace/source/glest_game/game/game_settings.h:186:30: warning: implicit conversion from 'float' to 'int' changes value from 1.5 to 1 [-Wliteral-conversion]
        fallbackCpuMultiplier = 1.5f;
                              ~ ^~~~
[ 73%] Building CXX object source/glest_game/CMakeFiles/megaglest.dir/gui/display.cpp.o
Segmentation fault (core dumped)
make[2]: *** [source/glest_game/CMakeFiles/megaglest.dir/build.make:401: source/glest_game/CMakeFiles/megaglest.dir/gui/display.cpp.o] Error 139
make[2]: *** Waiting for unfinished jobs....
```

andy5995 commented:
These segfaults may be related to https://github.com/tonistiigi/binfmt/issues/215

andy5995 committed:
Use native arm64 runner

https://github.blog/changelog/2025-01-16-linux-arm64-hosted-runners-now-available-for-free-in-public-repositories-public-preview/

Jammyjamjamman committed:
New workflows for the AppImage

Also add comment to explain changes to CmakeLists.txt
